Feature/data domain screen2

kotrasatheesh-dev opened the pull request.

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
Added Multi Module presentation

kotrasatheesh-dev committed:
feat:screen 2 data domain layers

kotrasatheesh-dev committed:
Feat:data and domain layers for screen2
1:domain and data/build.gradle.kts
2:CharacterDetailsMapper.kt,Character,Characterdetailsrepository,Characterdetailsrepositoryimpl,CharacterdetailsUsecase

kotrasatheesh-dev committed:
Feat:presentation screen2

kotrasatheesh-dev committed:
Fix:some changes made

kotrasatheesh-dev committed:
Delete presentation/build.gradle.kts

kotrasatheesh-dev committed:
Delete presentation/src/main/java/com/example/presentation/navigation/characters/AllCharacters.kt

kotrasatheesh-dev committed:
Fix:files some issues fixed

kotrasatheesh-dev committed:
Fix:files some issues fixed

kotrasatheesh-dev committed:
Merge branch 'feature/multi_module_data_domain' into feature/multi_module_presentation

# Conflicts:
#	app/build.gradle.kts
#	gradle/libs.versions.toml

kotrasatheesh-dev committed:
Fix:files some issues fixed

kotrasatheesh-dev committed:
Fix:files some issues fixed

kotrasatheesh-dev committed:
Fix:added test cases in TestCharacterDetailsUseCase

kotrasatheesh-dev committed:
Merge branch 'feature/multi_module_presentation' into feature/data_domain_screen2

ds-griddynamics commented on data/src/main/graphql/com/example/graphql/charactersDetailsById.graphql:
```diff
@@ -0,0 +1,32 @@
+  }
+}
+
+fragment character on Character {
```
move this outside to separate class

ds-griddynamics commented on data/src/main/java/data/repository/CharacterDetailsRepositoryImpl.kt:
```diff
@@ -0,0 +1,47 @@
+            if (response.hasErrors()) {
+                Result.failure(Exception(response.errors?.joinToString { it.message }))
+            } else {
+                val characterDetails = response.data?.character?.character
+                var episodes = characterDetails?.episode?.map { episode ->
+                    Character.Episode(
+                        id = episode?.episode.orEmpty(),
+                        name = episode?.name.orEmpty(),
+                        airDate = episode?.air_date.orEmpty()
+                    )
+                }
+                characterDetails?.let {
+                    Result.success(
+                        CharacterDetailsMapper(
+                            id = it.id,
+                            name = it.name,
+                            image = it.image,
+                            status = it.status,
+                            species = it.species,
+                            gender = it.gender,
+                            originName = it.origin?.name,
+                            originDimension = it.origin?.dimension,
+                            locationName = it.location?.location?.name,
+                            locationDimension = it.location?.location?.dimension,
+                            episodes = episodes
+                        )
+                    )
+                } ?: Result.failure(Exception("Character details not found"))
```
You can create a mapper class where all this can be done

ds-griddynamics commented on domain/src/main/java/domain/mapper/CharacterDetailsMapper.kt:
```diff
@@ -0,0 +1,17 @@
+import domain.repository.Character
+
+data class CharacterDetailsMapper(
+    val id: String? = "",
+    val name: String? = "",
+    val image: String? = "",
+    val status: String? = "",
+    val species: String? = "",
+    val gender: String? = "",
+    val originName: String? = "",
+    val originDimension: String? = "",
+    val locationName: String? = "",
+    val locationDimension: String? = "",
```
If we are having empty values, why this accept nullable?

kotrasatheesh-dev replied:
in this like name, status, species,  are nullable (String?) because they might not always be available or present in the data response. While the default value is an empty string (""), I initially set these as nullable to allow for cases where a value may be missing entirely, which aligns with the intent to represent missing data with null. the need   changing the approach If the values should always be present (even if empty) I can make these fields non-nullable (String) and default them to empty strings ("") If nullable values  appropriate, I can default them to null instead, to clearly represent missing data

ds-griddynamics reviewed (commented):
Nice

kotrasatheesh-dev committed:
Fix:PR comments resolved
CharacterDetailsMapper
CharacterDetailsMapperFactory.kt
CharacterDetailsRepositoryImpl
CharacterFragment.graphql
charactersDetailsById.graphql
